torchgen/api/cpp.py

```diff
@@ -248,33 +250,43 @@ def returntype_type(t: Type, *, mutable: bool, symint: bool = False) -> CType:
         elif t.name == BaseTy.Scalar:
             return BaseCType(scalarT)
     elif isinstance(t, ListType):
-        assert (
-            not mutable
-        ), "Native functions should never return a mutable tensor list. They should return void."
-        elem = returntype_type(t.elem, mutable=False)
+        if not is_foreach:
+            assert (
+                not mutable
+            ), "Native functions should never return a mutable tensor list. They should return void."
+        elem = returntype_type(t.elem, mutable=False, is_foreach=is_foreach)
```

Comment:
I'm not sure we want to special case return_type for foreach. It seems better to either (1) support returning multiple tensor list in general in the dispatcher, or (2) avoiding the dispatcher from having to be aware of this type of schema at all with some kind of special logic at the python level

(2) seems the easier choice. The concern is whether we are okay with having the python schema diverge from aten.  Technically, we already DO have a precedence of python schema diverging from aten schema in factory functions (but that is not super great - there is additional complexity to implement this if we want the cpp API to have parity, and people have been confused about where requires_grad-ness is being handled).

For (1), it appears from @bdhirsh's comment that there is some trickiness to doing this, https://github.com/pytorch/pytorch/pull/76126#discussion_r854570636, and we avoided doing it because method chaining was not a thing for those out= ops anyway.

I'm curious what @ezyang and @bdhirsh think of this. Given that we do care about method chaining in the case of foreach, maybe we should consider supporting functions returning mutable tensor list.

Reply:
I read the issue. @crcrpar, I'd love to understand what you're going for better, because I am not really seeing the benefit for this. As you say in the issue, this isn't enough to actually do method chaining. If you do want the list to be returned in public API, you can add a little wrapper function that re-returns the inputs. So I am puzzled what the value add here is.

Reply:
One motivation is this comment https://github.com/pytorch/pytorch/pull/104780#discussion_r1256289461. I thought it was okay as I haven't seen strong NO in https://github.com/pytorch/pytorch/issues/104817.

Reply:
Oh, I see, it's not for chaining, it's for renaming the variable as its meaning changes. Hmm, I guess I can't find too much fault in that.

Reply:
Returning to @soulitzer's review, I think I would prefer this to be solely a special case in Python binding generation, that only works for the codegenned bindings, and everything else is kept the same. This would substantially reduce the blast radius of this PR, since you no longer have to update all of the C++ implementations.